### ISSUE TITLE
[🐛 Bug]: move target out of bounds in chrome@107, wdio@8

### ISSUE THREAD

**Issue by DudaGod** [webdriverio/webdriverio#9620]
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md#reporting-new-issues).

### WebdriverIO Version

8.1.3

### Node.js Version

16.16.0

### Mode

Standalone Mode

### Which capabilities are you using?

_No response_

### What happened?

I got an error "move target out of bounds" when try to move cursor in left top corner of body for correct screen webview (without hover effect on any block). Before move cursor I called `scrollIntoView` and my test breaks.

### What is your expected behavior?

I expected that this code will works the same as with wdio@7.

### How to reproduce the bug.

Example of test:
```
const webdriverio = require('webdriverio');

(async () => {
    const browser = await webdriverio.remote({
        protocol: 'http',
        hostname: 'localhost',
        path: '/',
        port: 4444,
        automationProtocol: 'webdriver',
        logLevel: 'trace',
        connectionRetryCount: 0,
        capabilities: {
            browserName: 'chrome',
            browserVersion: '107.0',
        }
    });

    await browser.url('https://wikipedia.org');

    const body = await browser.$('body');

    await body.scrollIntoView();
    await body.moveTo({xOffset: 0, yOffset: 0});
})();
```

### Relevant log output

```typescript
❯❯❯ node wdio-test.js                                                                                                                                                                                                         ✘ 1
2023-01-18T13:08:17.960Z INFO webdriver: Initiate new session using the WebDriver protocol
2023-01-18T13:08:18.087Z INFO webdriver: [POST] http://localhost:4444/session
2023-01-18T13:08:18.087Z INFO webdriver: DATA {
  capabilities: {
    alwaysMatch: { browserName: 'chrome', browserVersion: '107.0' },
    firstMatch: [ {} ]
  },
  desiredCapabilities: { browserName: 'chrome', browserVersion: '107.0' }
}
2023-01-18T13:08:18.396Z INFO webdriver: COMMAND navigateTo("https://wikipedia.org/")
2023-01-18T13:08:18.397Z INFO webdriver: [POST] http://localhost:4444/session/9fd10be363ec038d8f265fcca2bd04b7/url
2023-01-18T13:08:18.397Z INFO webdriver: DATA { url: 'https://wikipedia.org/' }
2023-01-18T13:08:19.040Z INFO webdriver: RESULT null
2023-01-18T13:08:19.042Z INFO webdriver: COMMAND findElement("css selector", "body")
2023-01-18T13:08:19.043Z INFO webdriver: [POST] http://localhost:4444/session/9fd10be363ec038d8f265fcca2bd04b7/element
2023-01-18T13:08:19.043Z INFO webdriver: DATA { using: 'css selector', value: 'body' }
2023-01-18T13:08:19.096Z INFO webdriver: RESULT {
  'element-6066-11e4-a52e-4f735466cecf': 'de1c52ef-2444-44a0-a0d3-abaf64686eea'
}
2023-01-18T13:08:19.103Z INFO webdriver: COMMAND findElement("css selector", "html")
2023-01-18T13:08:19.103Z INFO webdriver: [POST] http://localhost:4444/session/9fd10be363ec038d8f265fcca2bd04b7/element
2023-01-18T13:08:19.103Z INFO webdriver: DATA { using: 'css selector', value: 'html' }
2023-01-18T13:08:19.154Z INFO webdriver: RESULT {
  'element-6066-11e4-a52e-4f735466cecf': '8f856b2b-0485-4318-bad4-0ec682f4d6d2'
}
2023-01-18T13:08:19.157Z INFO webdriver: COMMAND getElementRect("8f856b2b-0485-4318-bad4-0ec682f4d6d2")
2023-01-18T13:08:19.158Z INFO webdriver: [GET] http://localhost:4444/session/9fd10be363ec038d8f265fcca2bd04b7/element/8f856b2b-0485-4318-bad4-0ec682f4d6d2/rect
2023-01-18T13:08:19.205Z INFO webdriver: RESULT { height: 1145, width: 922, x: 0, y: 0 }
2023-01-18T13:08:19.206Z INFO webdriver: COMMAND executeScript(<fn>, <object>)
2023-01-18T13:08:19.207Z INFO webdriver: [POST] http://localhost:4444/session/9fd10be363ec038d8f265fcca2bd04b7/execute/sync
2023-01-18T13:08:19.207Z INFO webdriver: DATA {
  script: 'return (function (el) {\n' +
    '            if (!el || !el.getBoundingClientRect) {\n' +
    '                return;\n' +
    '            }\n' +
    '            const { left, top, width, height } = el.getBoundingClientRect();\n' +
    '            return {\n' +
    '                x: left + this.scrollX,\n' +
    '                y: top + this.scrollY,\n' +
    '                width,\n' +
    '                height\n' +
    '            };\n' +
    '        }).apply(null, arguments)',
  args: [
    {
      'element-6066-11e4-a52e-4f735466cecf': '8f856b2b-0485-4318-bad4-0ec682f4d6d2',
      ELEMENT: '8f856b2b-0485-4318-bad4-0ec682f4d6d2'
    }
  ]
}
2023-01-18T13:08:19.255Z INFO webdriver: RESULT { height: 1144.671875, width: 922, x: 0, y: 0 }
2023-01-18T13:08:19.256Z INFO webdriver: COMMAND getElementRect("de1c52ef-2444-44a0-a0d3-abaf64686eea")
2023-01-18T13:08:19.256Z INFO webdriver: [GET] http://localhost:4444/session/9fd10be363ec038d8f265fcca2bd04b7/element/de1c52ef-2444-44a0-a0d3-abaf64686eea/rect
2023-01-18T13:08:19.302Z INFO webdriver: RESULT { height: 1089, width: 922, x: 0, y: 40 }
2023-01-18T13:08:19.302Z INFO webdriver: COMMAND executeScript(<fn>, <object>)
2023-01-18T13:08:19.303Z INFO webdriver: [POST] http://localhost:4444/session/9fd10be363ec038d8f265fcca2bd04b7/execute/sync
2023-01-18T13:08:19.303Z INFO webdriver: DATA {
  script: 'return (function (el) {\n' +
    '            if (!el || !el.getBoundingClientRect) {\n' +
    '                return;\n' +
    '            }\n' +
    '            const { left, top, width, height } = el.getBoundingClientRect();\n' +
    '            return {\n' +
    '                x: left + this.scrollX,\n' +
    '                y: top + this.scrollY,\n' +
    '                width,\n' +
    '                height\n' +
    '            };\n' +
    '        }).apply(null, arguments)',
  args: [
    {
      'element-6066-11e4-a52e-4f735466cecf': 'de1c52ef-2444-44a0-a0d3-abaf64686eea',
      ELEMENT: 'de1c52ef-2444-44a0-a0d3-abaf64686eea'
    }
  ]
}
2023-01-18T13:08:19.347Z INFO webdriver: RESULT { height: 1088.671875, width: 922, x: 0, y: 40 }
2023-01-18T13:08:19.347Z INFO webdriver: COMMAND performActions(<object>)
2023-01-18T13:08:19.348Z INFO webdriver: [POST] http://localhost:4444/session/9fd10be363ec038d8f265fcca2bd04b7/actions
2023-01-18T13:08:19.348Z INFO webdriver: DATA {
  actions: [
    { id: 'action1', type: 'wheel', parameters: {}, actions: [Array] }
  ]
}
2023-01-18T13:08:19.598Z INFO webdriver: RESULT null
2023-01-18T13:08:19.599Z INFO webdriver: COMMAND releaseActions()
2023-01-18T13:08:19.599Z INFO webdriver: [DELETE] http://localhost:4444/session/9fd10be363ec038d8f265fcca2bd04b7/actions
2023-01-18T13:08:19.642Z INFO webdriver: RESULT null
2023-01-18T13:08:19.642Z INFO webdriver: COMMAND getElementRect("de1c52ef-2444-44a0-a0d3-abaf64686eea")
2023-01-18T13:08:19.643Z INFO webdriver: [GET] http://localhost:4444/session/9fd10be363ec038d8f265fcca2bd04b7/element/de1c52ef-2444-44a0-a0d3-abaf64686eea/rect
2023-01-18T13:08:19.688Z INFO webdriver: RESULT { height: 1089, width: 922, x: 0, y: 40 }
2023-01-18T13:08:19.689Z INFO webdriver: COMMAND executeScript(<fn>, <object>)
2023-01-18T13:08:19.689Z INFO webdriver: [POST] http://localhost:4444/session/9fd10be363ec038d8f265fcca2bd04b7/execute/sync
2023-01-18T13:08:19.689Z INFO webdriver: DATA {
  script: 'return (function (el) {\n' +
    '            if (!el || !el.getBoundingClientRect) {\n' +
    '                return;\n' +
    '            }\n' +
    '            const { left, top, width, height } = el.getBoundingClientRect();\n' +
    '            return {\n' +
    '                x: left + this.scrollX,\n' +
    '                y: top + this.scrollY,\n' +
    '                width,\n' +
    '                height\n' +
    '            };\n' +
    '        }).apply(null, arguments)',
  args: [
    {
      'element-6066-11e4-a52e-4f735466cecf': 'de1c52ef-2444-44a0-a0d3-abaf64686eea',
      ELEMENT: 'de1c52ef-2444-44a0-a0d3-abaf64686eea'
    }
  ]
}
2023-01-18T13:08:19.747Z INFO webdriver: RESULT { height: 1088.671875, width: 922, x: 0, y: 40 }
2023-01-18T13:08:19.747Z INFO webdriver: COMMAND executeScript(<fn>, <object>)
2023-01-18T13:08:19.747Z INFO webdriver: [POST] http://localhost:4444/session/9fd10be363ec038d8f265fcca2bd04b7/execute/sync
2023-01-18T13:08:19.748Z INFO webdriver: DATA {
  script: 'return (function () {\n' +
    '        return { scrollX: this.pageXOffset, scrollY: this.pageYOffset };\n' +
    '    }).apply(null, arguments)',
  args: []
}
2023-01-18T13:08:19.792Z INFO webdriver: RESULT { scrollX: 0, scrollY: 56 }
2023-01-18T13:08:19.793Z INFO webdriver: COMMAND performActions(<object>)
2023-01-18T13:08:19.793Z INFO webdriver: [POST] http://localhost:4444/session/9fd10be363ec038d8f265fcca2bd04b7/actions
2023-01-18T13:08:19.793Z INFO webdriver: DATA {
  actions: [
    {
      id: 'action2',
      type: 'pointer',
      parameters: [Object],
      actions: [Array]
    }
  ]
}
2023-01-18T13:08:19.843Z DEBUG webdriver: request failed due to response error: move target out of bounds
2023-01-18T13:08:19.843Z ERROR webdriver: Request failed with status 500 due to move target out of bounds: move target out of bounds
  (Session info: chrome=107.0.5304.87)
file:///my_project/node_modules/webdriver/build/utils.js:194
    return new CustomRequestError(body, requestOptions);
           ^

move target out of bounds: move target out of bounds
  (Session info: chrome=107.0.5304.87)
    at getErrorFromResponseBody (file:///my_project/node_modules/webdriver/build/utils.js:194:12)
    at NodeJSRequest._request (file:///my_project/node_modules/webdriver/build/request/index.js:164:23)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Browser.wrapCommandFn (file:///my_project/node_modules/webdriverio/node_modules/@wdio/utils/build/shim.js:72:29)
    at async PointerAction.perform (file:///my_project/node_modules/webdriverio/build/utils/actions/base.js:61:9)
    at async Element.wrapCommandFn (file:///my_project/node_modules/webdriverio/node_modules/@wdio/utils/build/shim.js:72:29)
    at async Element.elementErrorHandlerCallbackFn (file:///my_project/node_modules/webdriverio/build/middlewares.js:18:32)
    at async Element.wrapCommandFn (file:///my_project/node_modules/webdriverio/node_modules/@wdio/utils/build/shim.js:72:29)
    at async /my_project/wdio-test.js:23:5
```
```


### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Is there an existing issue for this?

- [X] I have searched the existing issues

**Comment by christian-bromann** [comment]
@DudaGod this is an error raised by Chromedriver due to miss usage of the actions API. I don't know why it is caused, it would require some investigations. Any contributions are welcome!

**Comment by sankalpguptasymphony** [comment]
@DudaGod I am just curious why it's trying to search css selector 'html'.
And can you please give it a try by mentioning //body instead of body.

![image](https://user-images.githubusercontent.com/99712194/213676293-98d6bad8-044b-41d6-9e09-e8693df4b646.png)

**Comment by DudaGod** [comment]
> I am just curious why it's trying to search css selector 'html'.

I don't know, with wdio@7 I don't see this kind of behavior.

> And can you please give it a try by mentioning //body instead of body.

It crashed my browser:

```
❯❯❯ node ./wdio-test.js                                                                                                                                                                                                       ✘ 1
2023-01-20T12:15:15.214Z INFO webdriver: Initiate new session using the WebDriver protocol
2023-01-20T12:15:15.359Z INFO webdriver: [POST] http://localhost:4444/session
2023-01-20T12:15:15.360Z INFO webdriver: DATA {
  capabilities: {
    alwaysMatch: { browserName: 'chrome', browserVersion: '107.0' },
    firstMatch: [ {} ]
  },
  desiredCapabilities: { browserName: 'chrome', browserVersion: '107.0' }
}
2023-01-20T12:15:15.658Z INFO webdriver: COMMAND navigateTo("https://wikipedia.org/")
2023-01-20T12:15:15.659Z INFO webdriver: [POST] http://localhost:4444/session/3725b41346c44ffc9228415a37281480/url
2023-01-20T12:15:15.659Z INFO webdriver: DATA { url: 'https://wikipedia.org/' }
2023-01-20T12:15:16.301Z INFO webdriver: RESULT null
2023-01-20T12:15:16.302Z INFO webdriver: COMMAND findElement("xpath", "//body")
2023-01-20T12:15:16.303Z INFO webdriver: [POST] http://localhost:4444/session/3725b41346c44ffc9228415a37281480/element
2023-01-20T12:15:16.303Z INFO webdriver: DATA { using: 'xpath', value: '//body' }
2023-01-20T12:15:16.494Z DEBUG webdriver: request failed due to response error: unknown error
2023-01-20T12:15:16.494Z ERROR webdriver: Request failed with status 500 due to unknown error: unknown error: session deleted because of page crash
from unknown error: cannot determine loading status
from tab crashed
  (Session info: chrome=107.0.5304.87)
file:///my_project/node_modules/webdriver/build/utils.js:194
    return new CustomRequestError(body, requestOptions);
           ^

unknown error: unknown error: session deleted because of page crash
from unknown error: cannot determine loading status
from tab crashed
  (Session info: chrome=107.0.5304.87)
    at getErrorFromResponseBody (file:///my_project/node_modules/webdriver/build/utils.js:194:12)
    at NodeJSRequest._request (file:///my_project/node_modules/webdriver/build/request/index.js:164:23)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Browser.wrapCommandFn (file:///my_project/node_modules/webdriverio/node_modules/@wdio/utils/build/shim.js:72:29)
    at async Browser.$ (file:///my_project/node_modules/webdriverio/build/commands/browser/$.js:89:17)
    at async Browser.wrapCommandFn (file:///my_project/node_modules/webdriverio/node_modules/@wdio/utils/build/shim.js:72:29)
    at async /my_project/wdio-test.js:20:18
```

But using `document.evaluate('//body', document, null, XPathResult.ANY_TYPE, null)` I can find `//body` on the page.

**Comment by sankalpguptasymphony** [comment]
@DudaGod Can you please update your capabilities as below and retest once -

![image](https://user-images.githubusercontent.com/99712194/213701801-033032e8-07af-4da3-95bc-9f50efade061.png)

**Comment by erwinheitzman** [comment]
If you add some pauses between your actions you will clearly see what is happening. Pair this with our [documentation](https://webdriver.io/docs/api/element/moveTo/) and it should be clear as to why you are getting this error.

Explanation:
- first you scroll the body into view which scrolls to the center of the body
- next you want to moveTo the 0, 0 positions which are now scrolled outside of the range of the viewport

I think the following change might cause this but it would have got nothing to do with v8: https://github.com/webdriverio/webdriverio/commit/ed9362080661c527242f88ec4135cbef18d2f499#diff-6cd658472952f924b0abda2509003ba66379c52a6079e9fe1719786fde92933a

**Comment by christian-bromann** [comment]
Thanks for looking into this @erwinheitzman 

> However it does not seem like this is the case. Is this assumption correct?

The `scrollIntoView` command should behave the same as the Web API. Can you spot any differences?

**Comment by DudaGod** [comment]
> first you scroll the body into view which scrolls to the center of the body (if not the center, it at least scrolls a bit)

But why `scrollIntoView` starts scroll the element to the center and not for the top like it was in wdio@7? Moreover `scrollIntoView` in browser scrolls to the top. So this behavior looks very strange.

> Christian, reading the code I believe that by default it should scroll the element to the top of the viewport.

+1

> The scrollIntoView command should behave the same as the Web API. Can you spot any differences?

Yes. `scrollIntoView` in browser move viewport to the top of the element, but wdio@8 doesn't.

**Comment by christian-bromann** [comment]
> But why `scrollIntoView` starts scroll the element to the center and not for the top like it was in wdio@7? Moreover `scrollIntoView` in browser scrolls to the top. So this behavior looks very strange.

We changed the underlying implementation to make it more consistent with the browser. That seems to be not 100% correct. Any contributions that can resolve the issue are appreciated.

**Comment by christian-bromann** [comment]
@DudaGod did you had any chance to look into this issue?

**Comment by DudaGod** [comment]
> @DudaGod did you had any chance to look into this issue?

I will try to fix it today.

**Comment by DudaGod** [comment]
I have a few questions after reviewing the source code:
- why we don't use simple `scrollIntoViewWeb` here - https://github.com/webdriverio/webdriverio/blob/main/packages/webdriverio/src/commands/element/scrollIntoView.ts#L44-L91 ? Which will correctly work with webdriver and devtools protocol
- i can't find `performAction` and `releaseAction` commands in webdriver protocol. They are called here - https://github.com/webdriverio/webdriverio/blob/main/packages/webdriverio/src/utils/actions/base.ts#L83-L86 and in IDE I see that these commands do not exist in `WebdriverIO.Browser` type
- `y` position inside `scrollIntoView` calculated not completely correctly - https://github.com/webdriverio/webdriverio/blob/main/packages/webdriverio/src/commands/element/scrollIntoView.ts#L69 because does not take into account the current position of the viewport. But in devtools protocol everything works fine since it is taken into account within the `performActions` command - https://github.com/webdriverio/webdriverio/blob/main/packages/devtools/src/commands/performActions.ts#L195-L217. Where the same code for webdriver protocol? (i don't see it)

**Comment by christian-bromann** [comment]
> * why we don't use simple `scrollIntoViewWeb` here

Because this would use web APIs on which we can't rely on given we never know the environment the test is executed in. The WebDriver action API is well defined and works across browser so there is no reason not to use it.

> * i can't find `performAction` and `releaseAction` commands in webdriver protocol.

These are protocol commands and are defined in [`@wdio/protocols`](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-protocols/src/protocols/webdriver.ts#L1136-L1159). Their types and functions are auto-generated like all other WebDriver commands. I am not sure why your IDE doesn't detect them but they are there.

> * because does not take into account the current position of the viewport

Maybe this is the but that needs to be fixed.

> * But in devtools protocol everything works fine since it is taken into account within the `performActions` command -

Not sure what you mean by that. The `webdriverio` package also uses `performActions` through the nicer [`action`](https://github.com/webdriverio/webdriverio/blob/main/packages/webdriverio/src/commands/element/scrollIntoView.ts#L81-L83) interface.

> * Where the same code for webdriver protocol? (i don't see it)

Here is the chain of commands:

1. call [`scrollIntoView`](https://github.com/webdriverio/webdriverio/blob/main/packages/webdriverio/src/commands/element/scrollIntoView.ts#L44)
2. call [`action`](https://github.com/webdriverio/webdriverio/blob/main/packages/webdriverio/src/commands/element/scrollIntoView.ts#L81)
3. create a [`WheelAction`](https://github.com/webdriverio/webdriverio/blob/main/packages/webdriverio/src/commands/browser/action.ts#L179)
4. call [`performAction`](https://github.com/webdriverio/webdriverio/blob/a32ecf2abb9202d532fb98b7c23ceed9dbb2f6ee/packages/webdriverio/src/utils/actions/base.ts#L83) when calling [`perform`](https://github.com/webdriverio/webdriverio/blob/a32ecf2abb9202d532fb98b7c23ceed9dbb2f6ee/packages/webdriverio/src/commands/element/scrollIntoView.ts#L83)

Let me know if you have any further questions.

**Comment by BorisOsipov** [comment]
>why we don't use simple scrollIntoViewWeb here
>Because this would use web APIs on which we can't rely on given we never know the environment the test is executed in. The WebDriver action API is well defined and works across browser so there is no reason not to use it.

To be honest, it is very strange. We have created a command related to elements which refers to browser-specific things. I mean, we refer to MDN in command docs, which is definitely web-specific, and then try to adopt and behave for all driver backends, which is not possible.
We probably have to remove or deprecate it instead of attempting to make it work on platforms that cannot do so.
I mean we can't make it work for Appium Android as well as the Web spec from MDN.

**Comment by BorisOsipov** [comment]
MDN is a doc we reffer to https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView

**Comment by christian-bromann** [comment]
This has been fixed so the following script is passing:
```ts
import { remote, Key } from 'webdriverio'

const browser = await remote({
    logLevel: 'error',
    capabilities: {
        browserName: 'chrome',
        'goog:chromeOptions': {
            // args: ['--headless=new']
        }
    }
})

try {
    await browser.url('https://wikipedia.org');

    const body = await browser.$('body');

    await body.scrollIntoView();
    await body.moveTo({xOffset: 0, yOffset: 0});

    await browser.deleteSession()
} catch (e) {
    console.log(e);
}
```

@BorisOsipov you are raising a valid concern. It would be much easier to just call `execute` and use Web APIs. Given that Appium won't support the wheel action type it seems less useful to have it implemented this way. The reason I switched to use the action API is because I wanted to more rely on the WebDriver protocol which ensures that browser automation is closer to human browser interaction. It is the same reason why we don't execute clicks though JS events. Doing so could probably resolve a lot of "element not intractable" issues however it would not be a "real" click. Whether using the wheel action makes it behave close like a user than calling the web API is a different question to find out but I hope this provides more context as to why this change happened.

I am happy adding a parameter allowing the user to use the web API instead.
